linera-core/src/client/mod.rs

```diff
@@ -2496,33 +2393,91 @@ where
             }
         }
 
-        // The block we want to propose is either the highest validated, or our pending one.
-        let Some(block) = manager
-            .highest_validated_block()
-            .cloned()
-            .or_else(|| self.state().pending_block().clone())
-        else {
-            return Ok(ClientOutcome::Committed(None)); // Nothing to propose.
+        // Otherwise we have to re-propose the highest validated block, if there is one.
+        let executed_block = if let Some(certificate) = &info.manager.requested_locked {
+            certificate.executed_block().clone()
+        } else {
+            let block = if let Some(proposal) = info
+                .manager
+                .requested_proposed
+                .as_ref()
+                .filter(|proposal| proposal.content.round.is_fast())
+            {
+                // The fast block counts as "locked", too. If there is one, re-propose that.
+                proposal.content.block.clone()
+            } else if let Some(block) = self.state().pending_block() {
+                block.clone() // Otherwise we are free to propose our own pending block.
+            } else {
+                return Ok(ClientOutcome::Committed(None)); // Nothing to do.
+            };
+            self.stage_block_execution(block).await?.0
         };
 
-        let round = match Self::round_for_new_proposal(&info, &identity, Some(&block))? {
-            Either::Left(round) => round,
-            Either::Right(timeout) => return Ok(ClientOutcome::WaitForTimeout(timeout)),
+        let round = match Self::round_for_new_proposal(
+            &info.manager,
+            &identity,
+            Some(&executed_block.block),
+        ) {
+            Ok(round) => round,
+            Err(error) => {
+                let timeout = info.round_timeout().ok_or(error)?;
```

Comment:
Why does `info.round_timeout()` take precedence here?

Reply:
Because `round_for_new_proposal` returns an error whenever we can't propose in the current round, and if we can make the current round time out then that's fine. But now I'm wondering why I moved creating the timeout out of that function. Will revert…